Running with --exclude='**' crashes

<!--

Welcome! - We kindly ask that you:

  1. Fill out the issue template below - not doing so needs a good reason.
  2. Use the forum if you have a question rather than a bug or feature request.

The forum is at: https://forum.restic.net

NOTE: Not filling out the issue template needs a good reason, as otherwise it
may take a lot longer to find the problem, not to mention it can take up a lot
more time which can otherwise be spent on development. Please also take the
time to help us debug the issue by collecting relevant information, even if
it doesn't seem to be relevant to you. Thanks!

The forum is a better place for questions about restic or general suggestions
and topics, e.g. usage or documentation questions! This issue tracker is mainly
for tracking bugs and feature requests directly relating to the development of
the software itself, rather than the project.

Thanks for understanding, and for contributing to the project!

-->
Thanks Tom (@tkw1536) for discovering this bug.

Output of `restic version`
--------------------------
`restic 0.12.0 compiled with go1.15.8 on linux/amd64`

How did you run restic exactly?
-------------------------------
```
[leonhard@red testing]$ restic -r . init
enter password for new repository: 
enter password again: 
created restic repository 9728a11fb9 at .

Please note that knowledge of your password is required to access
the repository. Losing your password means that your data is
irrecoverably lost.
[leonhard@red testing]$ restic -r . backup --exclude='**' /home/leonhard/dev/
enter password for repository: 
repository 9728a11f opened successfully, password is correct
created new cache in /home/leonhard/.cache/restic
no parent snapshot found, will read all files
panic: runtime error: index out of range [0] with length 0

goroutine 74 [running]:
github.com/restic/restic/internal/filter.match(0xc00020e9c0, 0x0, 0x4, 0xc0002242c0, 0x4, 0x4, 0x0, 0x3, 0x4)
	github.com/restic/restic/internal/filter/filter.go:173 +0x6f0
github.com/restic/restic/internal/filter.match(0xc0000ba060, 0x1, 0x1, 0xc0002242c0, 0x4, 0x4, 0x55960f7e3617, 0x0, 0x7ffe892c341a)
	github.com/restic/restic/internal/filter/filter.go:152 +0x58b
github.com/restic/restic/internal/filter.list(0xc0000ba080, 0x1, 0x1, 0x7ffe892c3400, 0x7ffe892c341a, 0x12, 0x55960ff97189, 0xc000038640, 0x1c)
	github.com/restic/restic/internal/filter/filter.go:241 +0x115
github.com/restic/restic/internal/filter.List(...)
	github.com/restic/restic/internal/filter/filter.go:221
main.rejectByPattern.func1(0x7ffe892c341a, 0x12, 0x0)
	github.com/restic/restic/cmd/restic/exclude.go:79 +0x5b
main.runBackup.func3(0x7ffe892c341a, 0x12, 0x0)
	github.com/restic/restic/cmd/restic/cmd_backup.go:612 +0x75
github.com/restic/restic/internal/archiver.(*Scanner).scan(0xc00015e7e0, 0x5596107e0800, 0xc000224000, 0x0, 0x0, 0x0, 0x0, 0x7ffe892c341a, 0x12, 0x1, ...)
	github.com/restic/restic/internal/archiver/scanner.go:107 +0x82
github.com/restic/restic/internal/archiver.(*Scanner).scanTree(0xc00015e7e0, 0x5596107e0800, 0xc000224000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x7ffe892c341a, 0x12, ...)
	github.com/restic/restic/internal/archiver/scanner.go:49 +0x13f
github.com/restic/restic/internal/archiver.(*Scanner).scanTree(0xc00015e7e0, 0x5596107e0800, 0xc000224000, 0x0, 0x0, 0x0, 0x0, 0xc000218bd0, 0x0, 0x0, ...)
	github.com/restic/restic/internal/archiver/scanner.go:60 +0x3d4
github.com/restic/restic/internal/archiver.(*Scanner).scanTree(0xc00015e7e0, 0x5596107e0800, 0xc000224000, 0x0, 0x0, 0x0, 0x0, 0xc000218ba0, 0x0, 0x0, ...)
	github.com/restic/restic/internal/archiver/scanner.go:60 +0x3d4
github.com/restic/restic/internal/archiver.(*Scanner).scanTree(0xc00015e7e0, 0x5596107e0800, 0xc000224000, 0x0, 0x0, 0x0, 0x0, 0xc000218b70, 0x0, 0x0, ...)
	github.com/restic/restic/internal/archiver/scanner.go:60 +0x3d4
github.com/restic/restic/internal/archiver.(*Scanner).Scan(0xc00015e7e0, 0x5596107e0800, 0xc000224000, 0xc0000a1b40, 0x1, 0x1, 0x55960f6b4d16, 0xc000218b48)
	github.com/restic/restic/internal/archiver/scanner.go:91 +0x17a
main.runBackup.func7(0x0, 0x0)
	github.com/restic/restic/cmd/restic/cmd_backup.go:671 +0x9c
gopkg.in/tomb%2ev2.(*Tomb).run(0xc0000c4550, 0xc0000d22c0)
	gopkg.in/tomb.v2@v2.0.0-20161208151619-d5d1b5820637/tomb.go:163 +0x3a
created by gopkg.in/tomb%2ev2.(*Tomb).Go
	gopkg.in/tomb.v2@v2.0.0-20161208151619-d5d1b5820637/tomb.go:159 +0xbc

```
<!--
This section should include at least:

 * The complete command line and any environment variables you used to
   configure restic's backend access. Make sure to replace sensitive values!

 * The output of the commands, what restic prints gives may give us much
   information to diagnose the problem!
-->

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
None, just a local folder as repository.


Expected behavior
-----------------
Backing up my folder `/home/leonhard/dev/`to the repository in `.` instead of crashing.
<!--
Describe what you'd like restic to do differently.
-->

Actual behavior
---------------
restic crashes. Please see the log above for details.
<!--
In this section, please try to concentrate on observations, so only describe
what you observed directly.
-->

Steps to reproduce the behavior
-------------------------------
1. Make a folder `mkdir testing`
2. Create a new restic repo `restic -r ./testing init`
3. Run backup including the option `restic -r ./testing backup --exclude='**' <some-other-folder>`

<!--
The more time you spend describing an easy way to reproduce the behavior (if
this is possible), the easier it is for the project developers to fix it!
-->

Do you have any idea what may have caused this?
-----------------------------------------------
Nope.


Do you have an idea how to solve the issue?
-------------------------------------------
You could hardcode this specific input to be ignored.


Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Absolutely, restic is great! I use it personally and professionally. It is a tool like no other for backups. Thank you for keeping it wel maintained!
<!--
Answering this question is not required, but if you have anything positive to share, please do so here!
Sometimes we get tired of reading bug reports all day and a little positive end note does wonders.
Idea by Joey Hess, https://joeyh.name/blog/entry/two_holiday_stories/
-->

--- a/changelog/unreleased/issue-3380
+++ b/changelog/unreleased/issue-3380
@@ -1,0 +1,7 @@
+Bugfix: Fix crash of `backup --exclude='**'`
+
+The exclude filter '**', which excludes all files, caused restic to crash. This
+has been fixed.
+
+https://github.com/restic/restic/issues/3380
+https://github.com/restic/restic/pull/3393

--- a/internal/filter/filter.go
+++ b/internal/filter/filter.go
@@ -166,6 +166,11 @@ func match(patterns Pattern, strs []string) (matched bool, err error) {
 		return true, nil
 	}
 
+	// an empty pattern never matches a non-empty path
+	if len(patterns) == 0 {
+		return false, nil
+	}
+
 	if len(patterns) <= len(strs) {
 		minOffset := 0
 		maxOffset := len(strs) - len(patterns)

--- a/internal/filter/filter_test.go
+++ b/internal/filter/filter_test.go
@@ -23,6 +23,7 @@ var matchTests = []struct {
 	{"*.go", "/foo/bar/test.go", true},
 	{"*.c", "/foo/bar/test.go", false},
 	{"*", "/foo/bar/test.go", true},
+	{"**", "/foo/bar/test.go", true},
 	{"foo*", "/foo/bar/test.go", true},
 	{"bar*", "/foo/bar/test.go", true},
 	{"/bar*", "/foo/bar/test.go", false},

